feature(cli): add vscode compatibility handler to inspector proxy

byCedric opened the pull request:
# Why

Builds upon #21449

This adds a handler that makes sure the `vscode-js-debug` extension can be used with the custom inspector proxy. These changes should NOT interfere with the other debuggers.

# How

- Added vscode compatibility handler
  - Add missing `description` to function nodes, as required by vscode-js-debug
    _(before https://github.com/microsoft/vscode-js-debug/issues/1583)_
  - Respond to `Debugger.getPossibleBreakpoints`, Hermes doesn't respond to this message
- ~~Filter out "useless" `callFrames` when debugging~~ 
  **Moved back to PR, we need to limit filtering**
  → https://github.com/expo/expo/pull/21663
- Avoid Hermes binding breakpoints to wrong locations, before sourcemap is loaded 
  → https://github.com/expo/expo/pull/21631
- Avoid Hermes crashing by disabling `Runtime.getProperties` on `symbol` types
  → https://github.com/expo/expo/pull/21639

# Test Plan

- See tests
- ~~Test project with vscode, see [this repository](https://github.com/byCedric/vscode-js-debug-issue-1583) for a test project.~~
- See https://github.com/expo/expo/pull/21641 to test `apps/bare-expo`

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

byCedric committed:
feature(cli): add vscode compatibility handler to inspector proxy

byCedric committed:
test(cli): add test for already existing descriptions

byCedric committed:
refactor(cli): integrate the vscode compat handler into inspector proxy

byCedric committed:
fix(cli): use better union types for vscode compatibility handler

byCedric committed:
docs(cli): add a note about hermes not being able to fetch function description

EvanBacon commented on packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/__tests__/VscodeCompat.test.ts:
```diff
@@ -0,0 +1,69 @@
+import { RuntimeGetProperties, VscodeCompatHandler } from '../VscodeCompat';
+
+describe(VscodeCompatHandler, () => {
```
top-level describe isn't needed and adds unneeded complexity to mock resetting.

byCedric replied:
Sure, I can remove it.

byCedric replied:
Let me remove it after I merged a few follow-up PRs like: #21566, else I'll be rebasing into oblivion

byCedric replied:
done

EvanBacon commented on packages/@expo/cli/src/start/server/metro/inspector-proxy/messages/VscodeCompat.ts:
```diff
@@ -0,0 +1,67 @@
+import type { Protocol } from 'devtools-protocol';
```
Assuming this is in devDependencies already?

byCedric replied:
Yep! See: https://github.com/expo/expo/pull/21449/files#diff-00b4bb0f6ea42a3d74201ea46359ab71c5085d6a17f80894dfc24d8614420018R135

byCedric committed:
chore(cli): add back the rebase-deleted network response tests

byCedric committed:
refactor(cli): drop describe from vscode compat test

byCedric committed:
docs(cli): add changelog entry

byCedric committed:
Update packages/@expo/cli/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

byCedric committed:
test(cli): update default handler count in test

byCedric committed:
fix(cli): filter hermes native callframes from `Debugger.paused` messages (#21566)

# Why

We don't need to show `callFrames` which are not traceable.

before | after
--- | ---
<img alt="image"
src="https://user-images.githubusercontent.com/1203991/223199631-fa5b5f25-733b-468d-b158-8e70b88c2b3d.png">
| <img alt="image"
src="https://user-images.githubusercontent.com/1203991/223199486-22496281-3de7-4693-a47b-63ec168762c8.png">


> ℹ️ **FunFact**
> The Chrome DevTools already filters these out from the `callFrames`

# How

- Filter `callFrames` in the `Debugger.paused` event

# Test Plan

- See added test

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

byCedric committed:
fix(cli): avoid hermes binding breakpoints to wrong locations using `setBreakpointByUrl` (#21631)

# Why

Vscode is eagerly trying to create breakpoints, even before the
sourcemap is properly loaded. It uses the
[`Debugger.setBreakpointyUrl`](https://chromedevtools.github.io/devtools-protocol/v8/Debugger/#method-setBreakpointByUrl),
with a `urlRegex`, which confuses Hermes.

> Vscode also incorrectly adds `file://` to the regex, already
containing a URL with protocol (e.g. `file://http://localhost/App.js`).
This is used to detect these specific breakpoint requests coming from
vscode.

My understanding is that Hermes is notifying the debugger of the loaded
bundle (including the sourcemap reference). After this notification, the
debuggers load the full sourcemap to understand the code that's being
executed. It looks like Hermes is using the `lineNumber` and
`columnNumber` from the actual bundle, resulting in `App.js:15` being
translated to `index.bundle:15`.

# How

This intercepts these invalid `urlRegex` from vscode, and asks Hermes to
create an unbounded breakpoint by avoiding `urlRegex`. After the
sourcemap is loaded in vscode, the unbounded breakpoint is reattempted
through `Debugger.setBreakpoint` using a `scriptId` that vscode resolves
using the proper `localRoot`/`remoteRoot` settings.

before | after
--- | ---

![vscode-breakpoint-worng-location](https://user-images.githubusercontent.com/1203991/224014318-251d65b4-7d57-4b4a-b5e3-ef40afb82e44.gif)
|
![vscode-breakpoint-correct-location](https://user-images.githubusercontent.com/1203991/224014649-8e219cb4-fea3-4871-984f-e632cdf0bce6.gif)


# Test Plan

- Clone https://github.com/byCedric/vscode-js-debug-issue-1583
- Set a breakpoint in `App.js` before connecting to Hermes
- The location of the breakpoint should not change

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

byCedric committed:
fix(cli): remove symbol objectIds to avoid vscode crashing hermes (#21639)

# Why

Remove `objectId` from symbol types to disable vscode to improperly
fetch object properties on symbols, causing Hermes to hard-crash.

vscode before | vsocde after
--- | ---

![vscode-symbol-hard-crash](https://user-images.githubusercontent.com/1203991/224037441-aba1da1f-e48b-43e7-b1ec-2e4d09ac8e2e.gif)
|
![vscode-symbol-no-crash](https://user-images.githubusercontent.com/1203991/224037191-fbbd88ed-e936-4008-8ce5-cb501e56eac5.gif)

chrome before | chrome after
--- | ---

![chrome-symbol-before](https://user-images.githubusercontent.com/1203991/224038387-0b360c7a-7a1e-4abf-b809-919f4b05f730.gif)
|
![chrome-symbol-after](https://user-images.githubusercontent.com/1203991/224038408-a843bb32-2541-441b-b7ad-e51ea6544fa1.gif)


# How

By removing the `objectId`, the vscode debugger doesn't have any
reference to request the `Runtime.getProperties` on the symbol itself.
This avoids being able to crash Hermes by running
`Runtime.getProperties` on `symbol` types.

# Test Plan

- Clone https://github.com/byCedric/vscode-js-debug-issue-1583
- Add `const testSymbol = Symbol('test');` near a breakpoint
- Try to expand the `testSymbol` in the variables explorer

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg:
updated a module plugin).

Kudo commented on packages/@expo/cli/src/start/server/metro/inspector-proxy/handlers/VscodeCompat.ts:
```diff
@@ -0,0 +1,121 @@
+    // Hermes adds traces of JSI to the callFrames, which are refering to native code.
+    // It doesn't make sense to show these to the user, so we filter them out.
+    if ('method' in message && message.method === 'Debugger.paused') {
+      message.params.callFrames = message.params.callFrames.filter(
+        (frame) =>
+          frame.location.scriptId !== HERMES_INVALID_SCRIPT_ID &&
+          frame.functionName !== HERMES_NATIVE_FUNCTION_NAME
+      );
```
would be good if we could keep at least one frame here for developers to know the crashes is somehow from native side.

byCedric replied:
Do you have an example or some code snippet that triggers the crash? If so, I can limit this filtering to breakpoint `callFrames` only, leaving the crash `callFrames` as is. Does that solve the concern?

byCedric replied:
It's not a required functionality to make it work, so for now I'll just disable this code to not-block the exception stacks.

byCedric committed:
fix(cli): force `Debugger.setBreakpointByUrl` to be unbounded

byCedric committed:
refactor(cli): disable `callFrame` filtering until we can do this for breakpoints only

byCedric committed:
refactor(cli): move `callFrame` filtering back to PR

byCedric committed:
test(cli): update `setBreakpointByUrl` tests

byCedric commented:
@Kudo @EvanBacon I'm merging this to main, with the exception of the `callFrame` filtering. See: https://github.com/expo/expo/pull/21560#discussion_r1131270155
